lsp-methods.el

```diff
@@ -329,37 +334,69 @@ interface TextDocumentItem {
 	      :version (lsp--cur-file-version)
 	      :text (buffer-substring-no-properties (point-min) (point-max))))))
 
-(defun lsp--shutdown-cur-workspace ()
-  "Shut down the language server process for ‘lsp--cur-workspace’."
-  (with-demoted-errors "LSP error: %S"
-    (lsp--send-request (lsp--make-request "shutdown" (make-hash-table)) t)
-    (lsp--send-notification (lsp--make-notification "exit" nil)))
-  (lsp--uninitialize-workspace))
-
 ;; Clean up the entire state of lsp mode when Emacs is killed, to get rid of any
 ;; pending language servers.
 (add-hook 'kill-emacs-hook #'lsp--global-teardown)
 
 (defun lsp--global-teardown ()
   (with-demoted-errors "Error in ‘lsp--global-teardown’: %S"
-    (maphash (lambda (_k value) (lsp--teardown-client value)) lsp--workspaces)))
+    (maphash (lambda (_k value) (lsp--teardown-workspace value)) lsp--workspaces)))
 
-(defun lsp--teardown-client (client)
-  (setq lsp--cur-workspace client)
+(defun lsp--teardown-workspace (workspace)
+  (setq lsp--cur-workspace workspace)
   (lsp--shutdown-cur-workspace))
 
+(defun lsp--shutdown-cur-workspace ()
+  "Shut down the language server process for ‘lsp--cur-workspace’."
+  (with-demoted-errors "LSP error: %S"
+    (lsp--send-request (lsp--make-request "shutdown" (make-hash-table)) t)
+    (lsp--send-notification (lsp--make-notification "exit" nil)))
+  (lsp--uninitialize-workspace))
+
 (defun lsp--uninitialize-workspace ()
   "When a workspace is shut down, by request or from just
 disappearing, unset all the variables related to it."
-  (remhash (lsp--workspace-root lsp--cur-workspace) lsp--workspaces)
-  (let (proc)
+  (let (proc
```

Comment:
Since you're touching this code anyway, you can refactor it by assigning `proc` while binding it (`(let ((proc (lsp-workspace-proc ...)) ...)` or something more fancy like `(pcase-exhaustive lsp--cur-workspace ((cl-struct lsp--workspace proc root) ...))`).
Also you can remove the `(with-current-buffer (current-buffer))`, it doesn't do anything.